service/tbc/notifier.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2025 Hemi Labs, Inc.
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package tbc
+
+import (
+	"context"
+	"crypto/rand"
+	"io"
+	"sync"
+)
+
+type notification string
+
+const (
+	NotificationBlock       notification = "block inserted"
+	NotificationBlockheader notification = "blockheader inserted"
+)
+
+type Notifier struct {
+	mtx sync.Mutex
+
+	listeners map[string]*Listener
+
+	// If true, the notifier will block on new messages until the queue for
+	// each listener's message channel is unblocked (i.e., read from).
+	// This should only be TRUE for test purposes.
+	blocking bool
+}
+
+type Listener struct {
+	id       string
+	notifier *Notifier
+	ch       chan notification
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (l *Listener) Unsubscribe() {
+	// Mark listener for deletion even if we block
+	// so we skip sending notifications to them
+	l.cancel()
+
+	l.notifier.mtx.Lock()
+	defer l.notifier.mtx.Unlock()
+
+	delete(l.notifier.listeners, l.id)
+}
+
+func (l *Listener) Listen() <-chan notification {
+	return l.ch
+}
+
+func NewNotifier(blocking bool) *Notifier {
+	n := Notifier{
+		listeners: make(map[string]*Listener),
+		blocking:  blocking,
+	}
+	return &n
+}
+
+func (n *Notifier) Subscribe(ctx context.Context, capacity uint64) (*Listener, error) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+
+	var nid [16]byte
+	for {
+		if _, err := io.ReadFull(rand.Reader, nid[:]); err != nil {
+			return nil, err
+		}
+		if _, ok := n.listeners[string(nid[:])]; !ok {
+			lctx, cancel := context.WithCancel(ctx)
+			l := Listener{
+				ch:       make(chan notification, capacity),
+				id:       string(nid[:]),
+				notifier: n,
+				ctx:      lctx,
+				cancel:   cancel,
+			}
+			n.listeners[string(nid[:])] = &l
+			return &l, nil
+		}
+	}
+}
+
+func (n *Notifier) Notify(ctx context.Context, message notification) error {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+
+	for _, l := range n.listeners {
```

Comment:
this is clever but if you have 10 blockers and the first is slow...

Maybe that was the intent all along but this needs documenting.